ols/src/cache/postgres_cache.py

```diff
@@ -1,18 +1,109 @@
+import logging
+import pickle
+from typing import Any, Optional
 
+import psycopg2
 from langchain_core.messages.base import BaseMessage
 
 from ols.app.models.config import PostgresConfig
 from ols.src.cache.cache import Cache
+from ols.src.cache.cache_error import CacheError
+
+logger = logging.getLogger(__name__)
 
 
 class PostgresCache(Cache):
-    """Cache that uses Postgres to store cached values."""
+    """Cache that uses Postgres to store cached values.
+
+    The cache itself is stored in following table:
+
+    ```
+         Column      |            Type             | Nullable | Default | Storage  |
+    -----------------+-----------------------------+----------+---------+----------+
+     user_id         | text                        | not null |         | extended |
+     conversation_id | text                        | not null |         | extended |
+     value           | bytea                       |          |         | extended |
+     updated_at      | timestamp without time zone |          |         | plain    |
+    Indexes:
+        "cache_pkey" PRIMARY KEY, btree (user_id, conversation_id)
+        "cache_key_key" UNIQUE CONSTRAINT, btree (key)
+        "timestamps" btree (updated_at)
+    Access method: heap
+    ```
+    """
+
+    CREATE_CACHE_TABLE = """
+        CREATE TABLE IF NOT EXISTS cache (
+            user_id         text NOT NULL,
+            conversation_id text NOT NULL,
+            value           bytea,
+            updated_at      timestamp,
+            PRIMARY KEY(user_id, conversation_id)
+        );
+        """
+
+    CREATE_INDEX = """
+        CREATE INDEX IF NOT EXISTS timestamps
+            ON cache (updated_at)
+        """
+
+    SELECT_CONVERSATION_HISTORY_STATEMENT = """
+        SELECT value
+          FROM cache
+         WHERE user_id=%s AND conversation_id=%s LIMIT 1
+        """
+
+    UPDATE_CONVERSATION_HISTORY_STATEMENT = """
+        UPDATE cache
+           SET value=%s, updated_at=CURRENT_TIMESTAMP
+         WHERE user_id=%s AND conversation_id=%s
+        """
+
+    INSERT_CONVERSATION_HISTORY_STATEMENT = """
+        INSERT INTO cache(user_id, conversation_id, value, updated_at)
+        VALUES (%s, %s, %s, CURRENT_TIMESTAMP)
+        """
+
+    DELETE_CONVERSATION_HISTORY_STATEMENT = """
+        DELETE FROM cache
+         WHERE (user_id, conversation_id) in
+               (SELECT user_id, conversation_id FROM cache ORDER BY updated_at LIMIT
+        """
+
+    QUERY_CACHE_SIZE = """
+        SELECT count(*) FROM cache;
+        """
 
     def __init__(self, config: PostgresConfig):
         """Create a new instance of Postgres cache."""
+        # initialize connection to DB
+        self.conn = psycopg2.connect(
+            host=config.host,
+            port=config.port,
+            user=config.user,
+            password=config.password,
+            dbname=config.dbname,
+            sslmode=config.ssl_mode,
+            sslcert=config.ca_cert_path,
```

Comment:
This param needs to be `sslrootcert=`. I have verified it with operator code.

Reply:
updated too